### DISCUSSION TITLE
(SQLAlchemy)AccessTokenDatabase is not cleaning up expired tokens (no logout)

### DISCUSSION THREAD

**Discussion by Hazedd** [fastapi-users/fastapi-users#857] [Ideas]
feature request, and it could be on the road-map already.

When using the database strategy and letting the login expire the stored token will never be deleted, this will fill up unnecessary database storage. I'm not using the other database adapters but i guess it will happen as well.

## Expected behavior
i.m.o. fastapi-users should have a garbage collector as a background task to cleanup expired tokens.

## Configuration
 - Python version : 3.10
 - FastAPI version : 0.70.1
 - FastAPI Users version : 9.2.0

### FastAPI Users configuration
SQLite / aiosqlite
fastapi-users[sqlalchemy2,oauth]

**Possible solution**:

SQLAlchemyAccessTokenDatabase class (and other database adapters) should have a garbage collect function that will remove expired tokens like:

```python
class SQLAlchemyAccessTokenDatabase():
    ...
    async def remove_expired_tokens(self, time_s: int):
        statement = delete(
            self.access_token_table,
            self.access_token_table.created_at <= (datetime.now(timezone.utc) - timedelta(0, time_s))
        )
        await self.session.execute(statement)
        await self.session.commit()
```
It can be used in the DatabaseStrategy where the lifetime_seconds is defined. *An other option could be to set a expiration datetime in the token database table so removing can be done from the SQLAlchemyAccessTokenDatabase store without knowing the lifetime_seconds.*

```python
class DatabaseStrategy():
    ...
    async def database_cleanup(self):
        await self.database.remove_expired_tokens(self.lifetime_seconds)
```
In my project i trigger the remove_expired_tokens within the SQLAlchemyAccessTokenDatabase.get_by_token(), it's a expensive operation but i solved my problem without adjusting a lot of code. 

To make sure fastapi-users is collecting garbage make it a repeated task in the auth router by using a decorator inspired by [fastapi-utils ](https://github.com/dmontagu/fastapi-utils/blob/master/fastapi_utils/tasks.py)

**Comment by frankie567** [comment]
Yes, I left that to the developer discretion on purpose. The problem here is that the solution we may come up with will be highly opinionated ; it'll probably need scheduler tasks, as you are showing in your proposal. IMO, this is out-of-scope of the project.

That said, it would be helpful to have a note about this in the documentation and provide some ideas to solve this.

**Reply by Hazedd** [reply]
Its on the border of the scope IMO: the project is providing a database solution and knowing most users will let the session expire without a logout the project should clean it. but like you say the solution will be opinionated and probably don't fit in all implementations of the project. 

I think a decent way to provide a easy cleanup is to store the token expiration datatime in the database so it can be cleaned without knowing it lifetime. I absolutely agree on the note about the known limitation. if you agree, i can make a pr in the next couple of days.
